src/charm.py

```diff
@@ -43,6 +44,77 @@ def __init__(self, *args: typing.Any) -> None:
         super().__init__(*args)
         self.framework.observe(self.on.config_changed, self.reconciliation)
         self.framework.observe(self.on.spring_boot_app_pebble_ready, self.reconciliation)
+        self.database: typing.Optional[DatabaseRequires] = self._setup_database_requirer(
+            "mysql", "mysql"
+        )
+
+    def _setup_database_requirer(self, relation_name: str, database_name: str) -> DatabaseRequires:
+        """Set up a DatabaseRequires instance.
+
+        The DatabaseRequires instance is an interface between the charm and various data providers.
+        It handles those relations and emit events to help us abstract these integrations.
+
+        Args:
+            relation_name: Name of the data relation
+            database_name: Name of the database (can be overwritten by the provider)
+
+        Returns:
+            DatabaseRequires object
+        """
+        database_requirer = DatabaseRequires(
+            self,
+            relation_name=relation_name,
+            database_name=database_name,
+        )
+        self.framework.observe(database_requirer.on.database_created, self._on_database_created)
+        self.framework.observe(self.on[relation_name].relation_broken, self._on_relation_broken)
+        return database_requirer
+
+    def _on_database_created(self, event: DatabaseCreatedEvent) -> None:
+        """Event triggered when a database was created for this application.
+
+        Args:
+            event: The DatabaseCreatedEvent object
+        """
+        logger.debug("Database credentials are received: %s", event.username)
+        self.reconciliation(event)
+
+    def _on_relation_broken(self, event: RelationBrokenEvent) -> None:
+        """Handle relation broken event.
+
+        Args:
+            event: The RelationBrokenEvent object
+        """
+        logger.debug("Relation with database removed: %s", event.relation.name)
+        self.reconciliation(event)
+
+    def _datasource(self) -> dict[str, str]:
+        """Compute datasource dict and return it.
+
+        Returns:
+            Datasource dict containing details about the data provider integration.
+        """
+        if self.database:
+            relations_data = list(self.database.fetch_relation_data().values())
+            if relations_data:
+                data = relations_data[0]
+                if "endpoints" in data and "username" in data and "password" in data:
```

Comment:
According to the schema below, all these fields are required. Do we really want to check they are present?

Reply:
Well that's better so we control the outcome and set everything to empty

Reply:
It seems inconsistent to me given we are not checking the presence of `endpoints`

Reply:
I think that this is a misunderstanding from my answer [here](https://github.com/canonical/spring-boot-k8s-operator/pull/8#discussion_r1098827937). Those keys are required as per the specification but I'm testing their presence here anyway (before using them).  

That said, it's true that I'm not checking that `endpoints` can be split.  
So I think that Arturo has a point.

Reply:
I mean, or we check everything, or we check nothing. I'm inclined to check. WDYT ?

Reply:
There is not mecanism in the relation that enforce schema concordance, therefore there could be an issue with the lib or the provider charm. I'd rather check as Niels stated.

We should log an warning, but we don't want to impact a potentially working workload, so behavior will always be to fallback to empty configuration.

As a rule of thumbs, every time we fallback to empty configuration a warning must be logged

Reply:
I revisited the checks